source/common/tcp_proxy/tcp_proxy.cc

```diff
@@ -124,53 +141,56 @@ Config::Config(const envoy::config::filter::network::tcp_proxy::v2::TcpProxy& co
   }
 }
 
-const std::string& Config::getRegularRouteFromEntries(Network::Connection& connection) {
+RouteConstSharedPtr Config::getRegularRouteFromEntries(Network::Connection& connection) {
   // First check if the per-connection state to see if we need to route to a pre-selected cluster
   if (connection.streamInfo().filterState().hasData<PerConnectionCluster>(
           PerConnectionCluster::key())) {
     const PerConnectionCluster& per_connection_cluster =
         connection.streamInfo().filterState().getDataReadOnly<PerConnectionCluster>(
             PerConnectionCluster::key());
-    return per_connection_cluster.value();
+
+    envoy::config::filter::network::tcp_proxy::v2::TcpProxy::DeprecatedV1::TCPRoute
+        per_connection_route;
+    per_connection_route.set_cluster(per_connection_cluster.value());
+    return std::make_shared<const TCPRoute>(*this, per_connection_route);
   }
 
-  for (const Config::Route& route : routes_) {
-    if (!route.source_port_ranges_.empty() &&
+  for (const TCPRouteConstSharedPtr& route : routes_) {
+    if (!route->source_port_ranges_.empty() &&
```

Comment:
Let's add a `bool matches(const Network::Connection&)` method to `Route` and lift this sequence of checks into the Route implementation struct and always-returns-false version to `WeightedClusterEntry`.

Then `routes_` can be a `std::vector<RouteConstSharedPtr>` and we can ditch `TCPRouteConstSharedPtr`.